File: CS/ManageUsersOnLogon.Module/DatabaseUpdate/Updater.cs
using System;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Updating;
using DevExpress.Persistent.Validation;
using DevExpress.ExpressApp.Security.Strategy;
using DevExpress.Persistent.Base.Security;
using DevExpress.Persistent.BaseImpl.PermissionPolicy;
using DevExpress.ExpressApp.Security;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Base;

namespace ManageUsersOnLogon.Module.DatabaseUpdate {
    public class Updater : ModuleUpdater {
        public Updater(IObjectSpace objectSpace, Version currentDBVersion) : base(objectSpace, currentDBVersion) { }
        public override void UpdateDatabaseAfterUpdateSchema() {
            base.UpdateDatabaseAfterUpdateSchema();
            #region Create Users for the Complex Security Strategy
            CreateUser(ObjectSpace, "Sam", "sam@example.com", string.Empty, true);
            CreateUser(ObjectSpace, "John", "john@example.com", string.Empty, false);
            ObjectSpace.CommitChanges();
            #endregion
        }
        public static IAuthenticationStandardUser CreateUser(IObjectSpace os, string userName, string email, string password, bool isAdministrator) {
            if (string.IsNullOrEmpty(userName) || string.IsNullOrEmpty(email)) {
                throw new ArgumentException("UserName and Email address are not specified!");
            }
            PermissionPolicyUser user = os.FindObject<PermissionPolicyUser>(new BinaryOperator("UserName", userName));
            if (user == null) {
                user = os.CreateObject<PermissionPolicyUser>();
                user.UserName = userName;
                user.IsActive = true;
                user.SetMemberValue("Email", email);
                user.SetPassword(password);
                PermissionPolicyRole role = isAdministrator ? GetAdministratorRole(os) : GetDefaultRole(os);
                user.Roles.Add(role);
                user.Save();
                if (Validator.RuleSet.ValidateTarget(os, user, DefaultContexts.Save).State == ValidationState.Valid) {
                    os.CommitChanges();
                }
            }
            return user;
        }
        public static PermissionPolicyRole GetAdministratorRole(IObjectSpace os) {
            PermissionPolicyRole administratorRole = os.FindObject<PermissionPolicyRole>(new BinaryOperator("Name", "Administrator"));
            if (administratorRole == null) {
                administratorRole = os.CreateObject<PermissionPolicyRole>();
                administratorRole.Name = "Administrator";
                administratorRole.IsAdministrative = true;
            }
            return administratorRole;
        }
        private static PermissionPolicyRole GetDefaultRole(IObjectSpace ObjectSpace) {
            PermissionPolicyRole defaultRole = ObjectSpace.FindObject<PermissionPolicyRole>(new BinaryOperator("Name", "Default"));
            if(defaultRole == null) {
                defaultRole = ObjectSpace.CreateObject<PermissionPolicyRole>();
                defaultRole.Name = "Default";

                defaultRole.AddObjectPermission<PermissionPolicyUser>(SecurityOperations.Read, "[Oid] = CurrentUserId()", SecurityPermissionState.Allow);
                defaultRole.AddNavigationPermission(@"Application/NavigationItems/Items/Default/Items/MyDetails", SecurityPermissionState.Allow);
                defaultRole.AddMemberPermission<PermissionPolicyUser>(SecurityOperations.Write, "ChangePasswordOnFirstLogon", "[Oid] = CurrentUserId()", SecurityPermissionState.Allow);
                defaultRole.AddMemberPermission<PermissionPolicyUser>(SecurityOperations.Write, "StoredPassword", "[Oid] = CurrentUserId()", SecurityPermissionState.Allow);
                defaultRole.AddTypePermissionsRecursively<PermissionPolicyRole>(SecurityOperations.Read, SecurityPermissionState.Deny);
                defaultRole.AddTypePermissionsRecursively<ModelDifference>(SecurityOperations.ReadWriteAccess, SecurityPermissionState.Allow);
                defaultRole.AddTypePermissionsRecursively<ModelDifferenceAspect>(SecurityOperations.ReadWriteAccess, SecurityPermissionState.Allow);
                defaultRole.AddTypePermissionsRecursively<ModelDifference>(SecurityOperations.Create, SecurityPermissionState.Allow);
                defaultRole.AddTypePermissionsRecursively<ModelDifferenceAspect>(SecurityOperations.Create, SecurityPermissionState.Allow);
            }
            return defaultRole;
        }
    }
}

File: VB/ManageUsersOnLogon.Module/DatabaseUpdate/Updater.vb
Imports System
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Updating
Imports DevExpress.Persistent.Validation
Imports DevExpress.ExpressApp.Security.Strategy
Imports DevExpress.Persistent.Base.Security
Imports DevExpress.Persistent.BaseImpl.PermissionPolicy
Imports DevExpress.ExpressApp.Security
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Base

Namespace ManageUsersOnLogon.Module.DatabaseUpdate
    Public Class Updater
        Inherits ModuleUpdater

        Public Sub New(ByVal objectSpace As IObjectSpace, ByVal currentDBVersion As Version)
            MyBase.New(objectSpace, currentDBVersion)
        End Sub
        Public Overrides Sub UpdateDatabaseAfterUpdateSchema()
            MyBase.UpdateDatabaseAfterUpdateSchema()
            '            #Region "Create Users for the Complex Security Strategy"
            CreateUser(ObjectSpace, "Sam", "sam@example.com", String.Empty, True)
            CreateUser(ObjectSpace, "John", "john@example.com", String.Empty, False)
            ObjectSpace.CommitChanges()
            '            #End Region
        End Sub
        Public Shared Function CreateUser(ByVal os As IObjectSpace, ByVal userName As String, ByVal email As String, ByVal password As String, ByVal isAdministrator As Boolean) As IAuthenticationStandardUser
            If String.IsNullOrEmpty(userName) OrElse String.IsNullOrEmpty(email) Then
                Throw New ArgumentException("UserName and Email address are not specified!")
            End If
            Dim user As PermissionPolicyUser = os.FindObject(Of PermissionPolicyUser)(New BinaryOperator("UserName", userName))
            If user Is Nothing Then
                user = os.CreateObject(Of PermissionPolicyUser)()
                user.UserName = userName
                user.IsActive = True
                user.SetMemberValue("Email", email)
                user.SetPassword(password)
                Dim role As PermissionPolicyRole = If(isAdministrator, GetAdministratorRole(os), GetDefaultRole(os))
                user.Roles.Add(role)
                user.Save()
                If Validator.RuleSet.ValidateTarget(os, user, DefaultContexts.Save).State = ValidationState.Valid Then
                    os.CommitChanges()
                End If
            End If
            Return user
        End Function
        Public Shared Function GetAdministratorRole(ByVal os As IObjectSpace) As PermissionPolicyRole
            Dim administratorRole As PermissionPolicyRole = os.FindObject(Of PermissionPolicyRole)(New BinaryOperator("Name", "Administrator"))
            If administratorRole Is Nothing Then
                administratorRole = os.CreateObject(Of PermissionPolicyRole)()
                administratorRole.Name = "Administrator"
                administratorRole.IsAdministrative = True
            End If
            Return administratorRole
        End Function

        Private Shared Function GetDefaultRole(ByVal ObjectSpace As IObjectSpace) As PermissionPolicyRole
            Dim defaultRole As PermissionPolicyRole = ObjectSpace.FindObject(Of PermissionPolicyRole)(New BinaryOperator("Name", "Default"))
            If defaultRole Is Nothing Then
                defaultRole = ObjectSpace.CreateObject(Of PermissionPolicyRole)()
                defaultRole.Name = "Default"
                defaultRole.AddObjectPermission(Of PermissionPolicyUser)(SecurityOperations.Read, "[Oid] = CurrentUserId()", SecurityPermissionState.Allow)
                defaultRole.AddNavigationPermission("Application/NavigationItems/Items/Default/Items/MyDetails", SecurityPermissionState.Allow)
                defaultRole.AddMemberPermission(Of PermissionPolicyUser)(SecurityOperations.Write, "ChangePasswordOnFirstLogon", "[Oid] = CurrentUserId()", SecurityPermissionState.Allow)
                defaultRole.AddMemberPermission(Of PermissionPolicyUser)(SecurityOperations.Write, "StoredPassword", "[Oid] = CurrentUserId()", SecurityPermissionState.Allow)
                defaultRole.AddTypePermissionsRecursively(Of PermissionPolicyRole)(SecurityOperations.Read, SecurityPermissionState.Deny)
                defaultRole.AddTypePermissionsRecursively(Of ModelDifference)(SecurityOperations.ReadWriteAccess, SecurityPermissionState.Allow)
                defaultRole.AddTypePermissionsRecursively(Of ModelDifferenceAspect)(SecurityOperations.ReadWriteAccess, SecurityPermissionState.Allow)
                defaultRole.AddTypePermissionsRecursively(Of ModelDifference)(SecurityOperations.Create, SecurityPermissionState.Allow)
                defaultRole.AddTypePermissionsRecursively(Of ModelDifferenceAspect)(SecurityOperations.Create, SecurityPermissionState.Allow)
            End If
            Return defaultRole
        End Function

    End Class
End Namespace